# Applying Fake-Authentic Classifier Over Facebook Political Ads

## Context:

## Problem:

## Question: 

## Methodology:


## Part 1: Learning a Classifier Model from Articles Review

In [1]:
!wget -c 'https://github.com/pitaconsumer/some-datasets/blob/master/572515_1037534_compressed_Fake.csv.zip?raw=true'
!wget -c "https://github.com/pitaconsumer/some-datasets/blob/master/572515_1037534_compressed_True.csv.zip?raw=true"
!unzip -o  "572515_1037534_compressed_Fake.csv.zip?raw=true"
!unzip -o "572515_1037534_compressed_True.csv.zip?raw=true"

--2020-07-17 16:16:51--  https://github.com/pitaconsumer/some-datasets/blob/master/572515_1037534_compressed_Fake.csv.zip?raw=true
Resolving github.com (github.com)... 192.30.255.112
Connecting to github.com (github.com)|192.30.255.112|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://github.com/pitaconsumer/some-datasets/raw/master/572515_1037534_compressed_Fake.csv.zip [following]
--2020-07-17 16:16:52--  https://github.com/pitaconsumer/some-datasets/raw/master/572515_1037534_compressed_Fake.csv.zip
Reusing existing connection to github.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/pitaconsumer/some-datasets/master/572515_1037534_compressed_Fake.csv.zip [following]
--2020-07-17 16:16:52--  https://raw.githubusercontent.com/pitaconsumer/some-datasets/master/572515_1037534_compressed_Fake.csv.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.192.133, 151.101.0.133,

In [21]:
!pip install spacy tqdm

In [3]:
!python -m spacy download en_core_web_lg

✔ Download and installation successful
You can now load the model via spacy.load('en_core_web_lg')


In [48]:
!pip install textblob

     |████████████████████████████████| 636 kB 13.5 MB/s eta 0:00:01


In [49]:
!pip install preprocessor

  Created wheel for preprocessor: filename=preprocessor-1.1.3-py3-none-any.whl size=4477 sha256=d2a8628697ae75b1cebe72d62cb55c5c5ff9e4d3958ef5edc1887e2f4cdbb65c
  Stored in directory: /root/.cache/pip/wheels/22/ae/f5/38dbc573cab3fc25d865fc55f8845d40742161416220532ea2
Successfully built preprocessor


In [51]:
!pip install seaborn

     |████████████████████████████████| 215 kB 342 kB/s eta 0:00:01


In [44]:
import spacy 
import en_core_web_lg

spacy.prefer_gpu()
nlp = en_core_web_lg.load()

In [53]:
import pandas as pd
import numpy as np
from tqdm import tqdm
tqdm.pandas(desc="progress bar")


import preprocessor
from textblob import TextBlob
import statistics
from typing import List


import scipy

import spacy
import nltk
import re

import matplotlib.pyplot as plt
%matplotlib inline 
import seaborn as sns

from sklearn.feature_extraction.text import TfidfVectorizer

In [8]:
fake_df = pd.read_csv('Fake.csv')

In [13]:
%time fake_df.shape

CPU times: user 26 µs, sys: 1e+03 ns, total: 27 µs
Wall time: 31.2 µs


(23481, 4)

In [7]:
true_df = pd.read_csv("True.csv")

In [26]:
fake_vectors = fake_df['text'].progress_apply(lambda x: pd.Series(nlp(x).doc.vector.tolist()))

progress bar:  63%|██████▎   | 14873/23481 [13:47<08:04, 17.78it/s]IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)

progress bar: 100%|██████████| 23481/23481 [22:56<00:00, 17.06it/s]


In [25]:
true_vectors = true_df['text'].progress_apply(lambda x: pd.Series(nlp(x).doc.vector.tolist()))

progress bar: 100%|██████████| 21417/21417 [17:33<00:00, 20.34it/s]


In [27]:
fake_vectors['y'] = 0
true_vectors['y'] = 1

In [28]:
fake_vectors.to_pickle('fake_vectors.pickle')
true_vectors.to_pickle('true_vectors.pickle')

## Part 2: Facebook Political Ads Classified Into Fake Versus Authentic Via Random Forest Model 

#### CLF from Training of fake data sets
#### Predict outcomes for Facebook 
    Y_fb_pred = clf.fit(X_train_fake, Y_train_true).predict(X_test_fb_pol)
clf.fit(X_fb_pol, Y_fb_pred).  #Test Set is FB Dataset

In [32]:
!pip install nltk

In [34]:
#from wordcloud import WordCloud,STOPWORDS
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer

import re,string,unicodedata
from nltk.stem import WordNetLemmatizer,PorterStemmer
import os
import gc
from nltk.tokenize import word_tokenize
from collections import  Counter

#stop = set(stopwords.words('english'))
#punctuation = list(string.punctuation)
#stop.update(punctuation)

In [59]:
fb = pd.read_csv('fbpac-ads-en-US.csv.xz') #'/Users/mehrunisaqayyum/Downloads/work/fbpac.csv'
fb

id  \
0       hyperfeed_story_id_5c9baa3ee0ec08073500042   
1       hyperfeed_story_id_5c9bb2a2413852086735771   
2       hyperfeed_story_id_5c9bb4fa461731e29426627   
3                                23843380741530360   
4       hyperfeed_story_id_5c9bb059454851c17741213   
...                                            ...   
162319                           23843108782710078   
162320                           23843034525850259   
162321                           23842997138670612   
162322  hyperfeed_story_id_5c8b16b11b8f86515960964   
162323                           23842885237930242   

                                                     html  political  \
0       <div class="_5pa- userContentWrapper"><div cla...          0   
1       <div class="_5pa- userContentWrapper"><div cla...          0   
2       <div class="_5pa- userContentWrapper"><div cla...          0   
3       <div class="_5pcr userContentWrapper"><div cla...          0   
4       <div class="_5pa- userContentWrapper"><div cla...          0   
...                                                   ...        ...   
162319  <div class="_5pcr userContentWrapper"><div cla...         12   
162320  <div class="_5pcr userContentWrapper"><div cla...          0   
162321  <div class="_5pcr userContentWrapper"><div cla...          0   
162322  <div class="_5pa- userContentWrapper"><div cla...          7   
162323  <div class="_5pcr userContentWrapper"><div cla...          0   

        not_political                                        title  \
0                   0                League of Conservation Voters   
1                   0                            Indivisible Guide   
2                   0               International Rescue Committee   
3                   0                 Covenant House International   
4                   1                           Planned Parenthood   
...               ...                                          ...   
162319              0                        Keep Them Accountable   
162320              0  National Republican Congressional Committee   
162321              0                              POW Action Fund   
162322              0                                Beto O'Rourke   
162323              0                                         ACLU   

                                                  message  \
0       <p>BREAKING: Trump’s Department of the Interio...   
1       <p>The Mueller investigation is over. Special ...   
2       <p>Zimbabwe is reeling from the impact of Cycl...   
3       <p>What more can you do in the final hours of ...   
4       <p>Say it loud, say it proud: Our rights, our ...   
...                                                   ...   
162319  <p>Rep. Katko voted for tax breaks for his wea...   
162320  <p>Illinois early voting is open NOW &amp; you...   
162321  <p>From your favorite peaks to the polling pla...   
162322  <p>Beto just announced he’s running for presid...   
162323  <p>Claim your FREE ACLU Voter sticker today to...   

                                                thumbnail  \
0       https://pp-facebook-ads.s3.amazonaws.com/v/t1....   
1       https://pp-facebook-ads.s3.amazonaws.com/v/t1....   
2       https://pp-facebook-ads.s3.amazonaws.com/v/t1....   
3       https://pp-facebook-ads.s3.amazonaws.com/v/t1....   
4       https://pp-facebook-ads.s3.amazonaws.com/v/t1....   
...                                                   ...   
162319  https://pp-facebook-ads.s3.amazonaws.com/v/t1....   
162320  https://pp-facebook-ads.s3.amazonaws.com/v/t1....   
162321  https://pp-facebook-ads.s3.amazonaws.com/v/t1....   
162322  https://pp-facebook-ads.s3.amazonaws.com/v/t1....   
162323  https://pp-facebook-ads.s3.amazonaws.com/v/t1....   

                           created_at                     updated_at   lang  \
0       2019-03-27 16:52:25.625455+00  2019-03-27 16:52:25.625455+00  en-US   
1       2019-03-27 17:28:14.096849+00  2019-03-27 17:28:14.096849+00 

In [37]:
fb.columns

Index(['id', 'html', 'political', 'not_political', 'title', 'message',
       'thumbnail', 'created_at', 'updated_at', 'lang', 'images',
       'impressions', 'political_probability', 'targeting', 'suppressed',
       'targets', 'advertiser', 'entities', 'page', 'lower_page', 'targetings',
       'paid_for_by', 'targetedness', 'listbuilding_fundraising_proba'],
      dtype='object')

In [38]:
fb.dtypes

id                                 object
html                               object
political                           int64
not_political                       int64
title                              object
message                            object
thumbnail                          object
created_at                         object
updated_at                         object
lang                               object
images                             object
impressions                         int64
political_probability             float64
targeting                          object
suppressed                         object
targets                            object
advertiser                         object
entities                           object
page                               object
lower_page                         object
targetings                         object
paid_for_by                        object
targetedness                      float64
listbuilding_fundraising_proba    

In [60]:
fe = ['title','message','paid_for_by']
text_fb = fb[fe]
text_fb.head(-10)

title  \
0                           League of Conservation Voters   
1                                       Indivisible Guide   
2                          International Rescue Committee   
3                            Covenant House International   
4                                      Planned Parenthood   
...                                                   ...   
162309              AFT - American Federation of Teachers   
162310                                             MoveOn   
162311  AAAS - The American Association for the Advanc...   
162312         Nathan Fletcher for County Supervisor 2018   
162313                                      Industry City   

                                                  message  \
0       <p>BREAKING: Trump’s Department of the Interio...   
1       <p>The Mueller investigation is over. Special ...   
2       <p>Zimbabwe is reeling from the impact of Cycl...   
3       <p>What more can you do in the final hours of ...   
4       <p>Say it loud, say it proud: Our rights, our ...   
...                                                   ...   
162309  <p>Mike DeWine was elected as Ohio’s attorney ...   
162310  <p>We resisted and won control of the House! N...   
162311  <p>Diversity in science, technology, and engin...   
162312  <p>As Supervisor, I will stand up to the polit...   
162313  <p>Support Industry City's plan to create 15,0...   

                                              paid_for_by  
0                           League of Conservation Voters  
1                                     Indivisible Project  
2                          International Rescue Committee  
3                            Covenant House International  
4                Planned Parenthood Federation of America  
...                                                   ...  
162309  AFT - American Federation of Teachers, not aut...  
162310  MoveOn.org Political Action, pol.moveon.org, n...  
162311                                                NaN  
162312                Nathan Fletcher for Supervisor 2018  
162313                                      Industry City  

[162314 rows x 3 columns]

### Text Cleaning
We will review text in columns 'title','message','paid_for_by' for our Natural Language Processing project.

In [61]:
'''Remove punctuation and "weird stuff like --" from ['title','message','paid_for_by'].'''

import re

def text_cleaner(text_fb):
    # Visual inspection identifies a form of punctuation spaCy does not
    # recognize: the double dash '--'.  Better get rid of it now!
    text = re.sub(r'--',' ',text_fb)
    text = re.sub("[\[].*?[\]]", "", text_fb)
    text = ' '.join(text_fb.split())
    return text

In [67]:
'''Make text lowercase, remove text in square brackets, remove punctuation and remove words containing numbers.'''
import re
import string

def clean_text_round1(text_fb):
    text = text_fb.title.str.lower()
    text = text_fb.message.str.lower()
    text = text_fb.paid_for_by.str.lower()
    text = re.sub('\[.*?\]', '', text_fb)
    text = re.sub('[%s]' % re.escape(string.punctuation), '', text_fb)
    text = re.sub('\w*\d\w*', '', text_fb)
    return text_fb

round1 = lambda x: clean_text_round1(x)

In [66]:
text_fb.title.str.lower()
#applied lower case method to title column
#how to iterate over every row?

0                       league of conservation voters
1                                   indivisible guide
2                      international rescue committee
3                        covenant house international
4                                  planned parenthood
                             ...                     
162319                          keep them accountable
162320    national republican congressional committee
162321                                pow action fund
162322                                  beto o'rourke
162323                                           aclu
Name: title, Length: 162324, dtype: object

In [68]:
#Example: alice = text_cleaner(alice)
text_fb_nlp = clean_text_round1(text_fb)

TypeError: expected string or bytes-like object

#### Methodology Note: 
We need to parse the cleaned data. The cleaned data represents the Facebook text in the 'message' column. We could include the 'title' column as well to match the 'title' with the corresponding message as a measure of checking authenticity the way the New York Times Challenge demonstrated.

In [69]:
# Parse the cleaned ads. This can take a bit.
nlp = spacy.load('en_core_web_lg')
fb_ad_doc = nlp(text_fb_nlp)  #doc = nlp("string")

#Iterate over tokens in a doc
for token in fb_ad_doc:
    print(token.text) #print (token.text)
    
    #Don't forget sentences_df

NameError: name 'text_fb_nlp' is not defined

### Importing second Facebook dataset to argue for business case.

In [ ]:
fb_likes = pd.read_csv('/Users/mehrunisaqayyum/Downloads/pseudo_facebook.csv')
fb_likes

In [ ]:
#Concatenating
fb_df2 = pd.concat([fb, fb_likes], ignore_index=True, sort =True)
fb_df2

In [ ]:
fb_df2.columns

In [ ]:
fb_df2.dtypes

### Feature Description: Second Facebook Dataset

ad_id is the id of specific ad set | Numeric

Reporting_start and reporting_end are the start and end dates of the each ad | Numeric

Campaign_id is the id assigned by the ad running company | Numeric- Negligible

fb_campaign_id is the id assigned by facebook for every ad set| Numeric- Negligible

age and gender talk about the demographics | Categorical

Interest1, Interest2, Interest3 are the user interests and likes of facebook users who were targeted for the ad | Categorical 

Impressions are the number of times the ad was shown to the users |Numeric

Clicks is the number of time users clicked on the ad | Numeric

spent is the amount of money spent on each campaign | Numeric

Totalconversions is the number of users who have clicked the ad and have made a purchase or installed an app
approved_conversions tells how many became actual active users | Numerica

### Observation: We have columns like "political", 'not_political', 'title' and 'message' of the advertisement; 'created at'; 'lang' for languages; 'political_probabilty', and 'paid_for_by'.

In [39]:
fb.describe()

political  not_political    impressions  political_probability  \
count  162324.000000  162324.000000  162324.000000          162324.000000   
mean        4.177663       0.794214       4.605049               0.928099   
std        10.574403       2.770883      40.689815               0.169513   
min         0.000000       0.000000       0.000000               0.000079   
25%         0.000000       0.000000       1.000000               0.951812   
50%         1.000000       0.000000       1.000000               0.997825   
75%         4.000000       1.000000       2.000000               0.999978   
max       488.000000     330.000000    3575.000000               1.000000   

        targetedness  listbuilding_fundraising_proba  
count  112747.000000                   152394.000000  
mean        3.939723                        0.545964  
std         1.758957                        0.333821  
min         0.000000                        0.000091  
25%         3.000000                        0.265190  
50%         4.000000                        0.423612  
75%         4.000000                        0.999893  
max        12.000000                        1.230622

### Descriptive Statistics: Average Dollars Spent on Facebook Ads
From another dataset.

In [ ]:
clicks_df = pd.read_csv('/Users/mehrunisaqayyum/Downloads/datasets_104115_247225_data.csv')
clicks_df

In [ ]:
clicks_df.columns

In [ ]:
#Used 'clicks' and 'spent' to calculate the cost per click and per ad.

click_count_per_add = clicks_df.clicks.sum()/clicks_df.clicks.count()
click_ratio = (clicks_df.spent.sum()/63)/clicks_df.clicks.count()
ad_spend_count_ratio = (clicks_df.spent.sum()/63)/clicks_df.clicks.count()
print('Total number of Ads purcheased by IRA from 2015 to 2017: ', clicks_df.clicks.count())
print('Total dollar spent by IRA from 2015 to 2017: ${:.2f}'.format(clicks_df.spent.sum()/63))
print('Average cost per Ad: ${:.2f}'.format(ad_spend_count_ratio))
print('Average number of clicks per Ad: {:.2f}'.format(click_count_per_add))
print('Avg cost per click: ${:.2f}'.format(click_ratio))

#### Observation
We see that we have 162,314 records of political ad data on Facebook. The last ten sample how ads were purchased by nonpartisan groups, like "League of Conservation Voters", political organizer groups, like "Indivisible Project", unions "AFT", and international nonprofits like "International Rescue Committee". 


### Parse the Text

### Pickle Data on Facebook ads

In [ ]:
# Let's pickle it for later use
data_df.to_pickle("corpus.pkl")

### Identify Stop Words

In [ ]:
stop=set(stopwords.words('english'))

def build_list(fb,col="title"):
    corpus=[]
    lem=WordNetLemmatizer()
    stop=set(stopwords.words('english'))
    new= fb[col].dropna().str.split()
    new=new.values.tolist()
    corpus=[lem.lemmatize(word.lower()) for i in new for word in i if(word) not in stop]
    
    return corpus

## Semantic Analysis: Turning FB message into Vectors
We note the stop words and review counts of words from tf-idf.

In [ ]:
# Turning FB message into Vectors
from sklearn.feature_extraction.text import TfidfVectorizer

X_train, X_test = train_test_split(emma_paras, test_size=0.4, random_state=0)

vectorizer = TfidfVectorizer(max_df=0.5, # drop words that occur in more than half the paragraphs
                             min_df=2, # only use words that appear at least twice
                             stop_words='english', 
                             lowercase=False, #don't convert everything to lower case (since proper names are people who are targeted in disinfo campaigns)
                             use_idf=True,#we definitely want to use inverse document frequencies in our weighting
                             norm=u'l2', #Applies a correction factor so that longer paragraphs and shorter paragraphs get treated equally
                             smooth_idf=True #Adds 1 to all document frequencies, as if an extra document existed that used every word once.  Prevents divide-by-zero errors
                            )


#Applying the vectorizer
fb_message_tfidf =vectorizer.fit_transform(text_fb)
print("Number of features: %d" % fb_message_tfidf.get_shape()[1])

#splitting into training and test sets
X_train_tfidf, X_test_tfidf= train_test_split(fb_message_tfidf, test_size=0.4, random_state=0)


#Reshapes the vectorizer output into something people can read
X_train_tfidf_csr = X_train_tfidf.tocsr()

#number of paragraphs
n = X_train_tfidf_csr.shape[0]
#A list of dictionaries, one per paragraph
tfidf_bypara = [{} for _ in range(0,n)]
#List of features
terms = vectorizer.get_feature_names()

print(terms)


In [ ]:
#for each paragraph, lists the feature words and their tf-idf scores
for i, j in zip(*X_train_tfidf_csr.nonzero()):
    tfidf_bypara[i][terms[j]] = X_train_tfidf_csr[i, j]

#Keep in mind that the log base 2 of 1 is 0, so a tf-idf score of 0 indicates that the word was present once in that sentence.
print('Original sentence:', X_train[5])
print('Tf_idf vector:', tfidf_bypara[5])

In [ ]:
corpus=build_list(text_fb)
counter=Counter(corpus)
most=counter.most_common()
x=[]
y=[]
for word,count in most[:20]:
    if (word not in stop) :
        x.append(word)
        y.append(count)

In [ ]:
plt.figure(figsize=(9,7))
sns.barplot(x=y,y=x)
plt.title("most common word in title")

#### Observations:
Top 5 most common words are ranked:
    1) 'Committee'
    2)'International'
    3) 'Action' and 'Planned'
    4) 'Parenthood'
    
USA and 'America' are interchangeable--so maybe these are double-counting. 

'Democratic' not used as much as nonpolitical word of 'rescue'. However, we do not see 'Republican' or 'GOP'. 

'Beto' and'O'Rourke' are occuring at same rate. This is the only person to show up in this political ad data set. 

### Most Common 'Paid For By'

In [ ]:
corpus=build_list(text_fb,"paid_for_by")
counter=Counter(corpus)
most=counter.most_common()
x=[]
y=[]
for word,count in most[:10]:
    if (word not in stop) :
        x.append(word)
        y.append(count)
        
plt.figure(figsize=(9,7))
sns.barplot(x=y,y=x)
plt.title("Most Common Word in 'paid_for_by'")

### Figure of 'Number of Top Ad Titles'
To highlight the ads that we would should review for fake versus true classifier to strenghten business case for impact.

In [ ]:
def plot_count(feature, title,fb, size=1, show_percents=False):
    f, ax = plt.subplots(1,1, figsize=(4*size, 4))
    total = float(len(fb))
    g = sns.countplot(fb[feature],order = fb[feature].value_counts().index[0:20], palette='Set3')
    g.set_title("Number of {}".format(title))
    if (size > 2):
        plt.xticks(rotation=90, size=10)
    if(show_percents):
        for p in ax.patches:
            height = p.get_height()
            ax.text(p.get_x() + p.get_width()/2.,
                   height + 3, '{:1.2f}%'.format(100*height/total),
                   ha="center")
    ax.set_xticklabels(ax.get_xticklabels());
    plt.show()

In [ ]:
plot_count('title','Top Ad Titles', text_fb, 3.5)
#plt.title("Number of Top Ad Titles")

### Figure: Number of Most Popular Messages

In [ ]:
#Note the number (counts) of message.
plot_count('message','message countplot', text_fb, 3.5)

## Bag of Words
Defining functions to identify most common words and the create features from those words in the text. We need to create a df of the "corpus".

In [ ]:
#We need to use fb_text_df.
 ##Do we need to create features from Bag of Words and vectorize them for the Random Forest Classifier?
def bag_of_words(text):
    allwords = [token.lemma_
               for token in text
               if not token.is_punct
               and not token.is_stop]
    return [item[0] for item in Counter(allwords).most_common(2000)]
print()


In [ ]:
def bow_features(sentences, common_words):
    fb_text_df = pd.DataFrame(columns=common_words)
    fb_text_df['text_sentence'] = sentences[0]
    fb_text_df['text_source'] = sentences[1]
    fb_text_df.loc[:,common_words] = 0
    
    for i, sentence in enumerate(df['text_sentence']):
        words = [token.lemma_ 
                for token in sentence
                if (
                    not token.is_punct
                    and not token.is_stop
                    and token.lemma_ in common_words
                )]
        for word in words:
            fb_text_df.loc[i, word] += 1
        if i%100 == 0:
            print('Processing row {}'.format(i))
    return fb_text_d

In [ ]:
#data means fb_text_df
top_words_in_ads_dict = {}
for c in fb_text_df.columns:
    top = fb_text_df[c].sort_values(ascending=False).head(30)
    top_words_in_ads_dict[c]= list(zip(top.index, top.values))

top_words_in_ads_dict

### Frequency of Words


In [ ]:
fb['message'].dtype

In [ ]:
fb.message.iloc[1:10]

In [29]:
temp_variable = "".join(fb.message.iloc[1:1000])

NameError: name 'fb' is not defined

### Convert sentences into numeric vectors.
We need to transform sentences into NUMERIC vectors so that the vectors can be included in a Random Forest Classifier model, which cannot use string values.

First we must create list.

In [ ]:
#List 

message_propoganda_list = []
for topic in [#Need to include topic_raw like physics_raw]:
    for sentence in topic['title','paid_for_by']['docs']: #What is docs?
        message_propoganda_list = message_propoganda_list + sentence['message']  #Used message_propaganda for abstract
        
  

In [ ]:
#Use spacy to count frequency of words

!python -m spacy download en_core_web_sm
nlp = spacy.load('en_core_web_sm')

# All the processing work is done here, so it may take a while.
FB_political_doc = nlp(temp_variable)
#FB_popuarity_doc = nlp()

In [ ]:
# Cleaned Text before Identifying 'Stopwords'
from collections import Counter

# Utility function to calculate how frequently words appear in the data sets.
def word_frequencies(text, include_stop=True):
    
    # Build a list of words.
    # Strip out punctuation and, optionally, stop words.
    words = []
    for token in text:
        if not token.is_punct and (not token.is_stop or include_stop):
            words.append(token.text)
            
    # Build and return a Counter object containing word counts.
    return Counter(words)
    
# The most frequent words:
political_freq = word_frequencies(temp_variable).most_common(10)
#like_freq = word_frequencies().most_common(10)
print('Political:', political_freq)
#print('Popular:', like_freq)

## Sentiment Analysis
Source: https://github.com/adashofdata/nlp-in-python-tutorial/blob/master/3-Sentiment-Analysis.ipynb Regarding text data, there are a few popular techniques that we'll be going through in the next few notebooks, starting with sentiment analysis. A few key points to note with sentiment analysis.

TextBlob Module: Linguistic researchers have labeled the sentiment of words based on their domain expertise. Sentiment of words can vary based on where it is in a sentence. The TextBlob module allows us to take advantage of these labels. Sentiment Labels: Each word in a corpus is labeled in terms of polarity and subjectivity (there are more labels as well, but we're going to ignore them for now). A corpus' sentiment is the average of these. Polarity: How positive or negative a word is. -1 is very negative. +1 is very positive. Subjectivity: How subjective, or opinionated a word is. 0 is fact. +1 is very much an opinion. For more info on how TextBlob coded up its sentiment function.

In [ ]:
fb_df2.loc[0:5, 'message']

In [ ]:
TextBlob('The Mueller investigation is over').sentiment


In [ ]:
# Create quick lambda functions to find the polarity and subjectivity of each message
# Terminal / Anaconda Navigator: conda install -c conda-forge textblob

pol = lambda x: TextBlob(x).sentiment.polarity
sub = lambda x: TextBlob(x).sentiment.subjectivity

fb_df2['polarity'] = fb_df2['message'].apply(pol)
fb_df2['subjectivity'] = fb_df2['title'].apply(sub)
fb_df2

In [ ]:
#Loop each 'message' value

for token in text